### PR TITLE
feat: get the number of close dht nodes with announce/store support

### PR DESCRIPTION
**Pull request by Green-Sky**
number of total and how many >= 0.2.18 peers you are connected with.

This change is [Reviewable](https://reviewable.io/reviews/toktok/c-toxcore/2357)

### CODE REVIEW

File: toxcore/tox_private.h [hunk 1]
```diff
@@ -139,6 +139,13 @@ typedef enum Tox_Err_Dht_Get_Nodes {
 bool tox_dht_get_nodes(const Tox *tox, const uint8_t *public_key, const char *ip, uint16_t port,
                        const uint8_t *target_public_key, Tox_Err_Dht_Get_Nodes *error);
 
+/**
+ * This function returns the ratio of close dht nodes that are known to support announce/store.
```
Comment (iphydf) [inline toxcore/tox_private.h:143]: What will the client do with this information?

File: toxcore/tox_private.h [hunk 2]
```diff
@@ -139,6 +139,13 @@ typedef enum Tox_Err_Dht_Get_Nodes {
...
+ *
+ * @return ratio
+ */
+float tox_dht_get_announce_capability_ratio(const Tox *tox);
```
Comment (iphydf) [inline toxcore/tox_private.h:147]: Can we return a uint between 0 and 100? We don't currently have floats in the tox (or toxav) API. I'd prefer not to add them.

**Review by iphydf — CHANGES_REQUESTED** [review]

File: toxcore/tox_private.h (hunk 1 above)
Comment (Green-Sky) [reply toxcore/tox_private.h:143]: This info is purely diagnostic, to get the ratio of connected DHT nodes that are >= `0.2.18`. Thats why it is in `tox_private.h`.
I am thinking of redesigning this, so you can get the raw numbers (total- & new-nodes).

Comment (iphydf) [reply toxcore/tox_private.h:143]: Raw numbers is better, I think. It avoids floats and gives additional information.

**Review by Green-Sky — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/toktok/c-toxcore/2357)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @iphydf)

*[a discussion](https://reviewable.io/reviews/toktok/c-toxcore/2357#gh-1390340051:-Nj4VkL7E8H2fj5Ti2yP:b-896fix) (no related file):*
<details><summary><i>Previously, iphydf wrote…</i></summary><blockquote>

Raw numbers is better, I think. It avoids floats and gives additional information.
</blockquote></details>

Done.

___
*[a discussion](https://reviewable.io/reviews/toktok/c-toxcore/2357#gh-1390340185:-Nj4VjeY3k-OVIJffI_z:b-896fix) (no related file):*
<details><summary><i>Previously, iphydf wrote…</i></summary><blockquote>

*(Reviewable was [unable](https://github.com/Reviewable/Reviewable/issues/14) to map [this GitHub inline comment thread](https://github.com/TokTok/c-toxcore/pull/2357#discussion_r1390340185) to the right spot — sorry!)*

Can we return a uint between 0 and 100? We don't currently have floats in the tox (or toxav) API. I'd prefer not to add them.
</blockquote></details>

Done.

___

File: toxcore/tox_private.c [hunk 3]
```diff
@@ -149,3 +149,20 @@ bool tox_dht_get_nodes(const Tox *tox, const uint8_t *public_key, const char *ip
 
     return true;
 }
+
+uint16_t tox_dht_get_num_closelist(const Tox *tox) {
+    tox_lock(tox);
+    uint16_t num_total = dht_get_num_closelist(tox->m->dht);
```
Comment (iphydf) [inline toxcore/tox_private.c:155]: const (this is why cimple fails)

**Review by iphydf — CHANGES_REQUESTED** [review]

File: toxcore/DHT.c [hunk 4]
```diff
@@ -2965,6 +2965,34 @@ bool dht_non_lan_connected(const DHT *dht)
...
+            ++num_valid_close_clients;
+        }
+    }
+
+    return num_valid_close_clients;
+}
+
+uint16_t dht_get_num_closelist_announce_capable(const DHT *dht) {
+    uint16_t num_valid_close_clients_with_cap = 0;
+    for (size_t i = 0; i < LCLIENT_LIST; ++i) {
```
Comment (iphydf) [inline toxcore/DHT.c:2984]: uint32_t, because that's what dht_get_close_client expects.

**Review by iphydf — CHANGES_REQUESTED** [review]

**Review by iphydf — APPROVED** [review]
LGTM after the int type fix.

File: toxcore/DHT.c (hunk 4 above)
Comment (Green-Sky) [reply toxcore/DHT.c:2984]: right

**Pull request merged**